Tutorial notebooks overhaul

richardarsenault opened the pull request:
## Overview

Following the comments received during the CSHS workshop, it was clear that the notebooks weren't as clear as they should have been to allow easier user onboarding. Therefore the structure has been remodeled and much more comments are provided to help the users interpret what they are doing and how to play with the models, the inputs and the outputs.

Changes:
Replaced the 9 first notebooks with 10 new notebooks, including notebook 00_Introduction_to_JupyterLab to iron-out some of the difficulties seen in managing the platform.

Also updated the notebook index for sphinx autodocs.

huard commented:
@cjauvin Could you review the code?
@Bijan55699 Could you go through the notebooks and comment on the clarity of explanations?

Thanks, no rush.

cjauvin commented:
@huard will do!

richardarsenault commented:
@cjauvin Thanks! There the notebooks take users by the hand to explore the functionalities and features better and more in-depth, so the code is not meant to be "developer-level" but "beginner-level", so please don't work your magic to optimize everything 😄 

Also, some notebooks need data that is generated in a previous notebook (kind of a daisy-chain type of thing), where we progressively increase the complexity. So auto-tests might fail if they are run independently.

Bijan55699 commented:
> @cjauvin Could you review the code? @Bijan55699 Could you go through the notebooks and comment on the clarity of explanations?
> 
> Thanks, no rush.

@huard : will do as well

tlvu commented on docs/source/notebooks/00_Introduction_to_JupyterLab.ipynb:
```diff
@@ -0,0 +1,242 @@
+{
```
<p>Add link to "PAVICS-Hydro documentation"?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

richardarsenault replied:
<p>Ideally yes, but I would need somebody to put the file on the website and send me the link to it.</p>

richardarsenault replied:
@huard @tlogan2000 The document is here, is it possible to host it somewhere so we can link to it in the notebooks?

[PAVICS-Hydro Tutorial.pdf](https://github.com/Ouranosinc/raven/files/8985726/PAVICS-Hydro.Tutorial.pdf)

thanks!

tlvu commented on docs/source/notebooks/00_Introduction_to_JupyterLab.ipynb:
```diff
@@ -0,0 +1,242 @@
+{
```
<p>you can simply add a cell and "pip install <strong>package</strong>" as required</p><p> </p><p>It should be "! pip install <strong>package"</strong> to execute code in a cell.</p><p> </p><p>However use can also use the jupyter conda plugin to install via conda (Settings -&gt; Conda Packages Manager).</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

richardarsenault replied:
<p>thanks, added!</p>

huard commented on docs/source/notebooks/05_Advanced_RavenPy_configuration.ipynb:
```diff
@@ -4,25 +4,20 @@
    "cell_type": "markdown",
```
<p>Check title formatting code (###)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

richardarsenault replied:
<p>Here it shows as "## #" but on the pavics server it is displayed normally?</p>

huard commented on docs/source/notebooks/05_Advanced_RavenPy_configuration.ipynb:
```diff
@@ -4,25 +4,20 @@
    "cell_type": "markdown",
```
<p>review markup</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

richardarsenault replied:
<p>Same here, on the server it looks perfect. Perhaps a problem with ReviewNB?</p>

huard commented on docs/source/notebooks/05_Advanced_RavenPy_configuration.ipynb:
```diff
@@ -4,25 +4,20 @@
    "cell_type": "markdown",
```
<pre class="ql-syntax" spellcheck="false">Line #4.    print(config)
</pre><p>I'd do something like</p><pre class="ql-syntax" spellcheck="false">print([f.name for f in config])
</pre><p>and skip the next cell. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

richardarsenault replied:
<p>Good call, changed.</p>

huard commented on docs/source/notebooks/06_Raven_calibration.ipynb:
```diff
@@ -15,26 +15,19 @@
    "source": [
```
<p>setting up</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

richardarsenault replied:
<p>thx</p>

huard commented on docs/source/notebooks/07_Making_and_using_hotstart_files.ipynb:
```diff
@@ -4,7 +4,7 @@
    "cell_type": "markdown",
```
<p>I think we should promote the "hru" formulation instead of passing area, elevation, latitude and longitude directly.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

richardarsenault replied:
<p>True, good catch. fixed.</p>

huard commented on docs/source/notebooks/07_Making_and_using_hotstart_files.ipynb:
```diff
@@ -4,7 +4,7 @@
    "cell_type": "markdown",
```
<p>The following notebooks do not use this.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

richardarsenault replied:
<p>Actually we do in notebook 12, but I specified it directly to clarify.</p>

huard commented on docs/source/notebooks/08_Getting_and_bias_correcting_CMIP6_data.ipynb:
```diff
@@ -0,0 +1,1788 @@
+{
```
<p>Nitpicking, but I think end_day would be closer to what you mean than stop_day.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

richardarsenault replied:
<p>Good point, changed.</p>

huard commented on docs/source/notebooks/08_Getting_and_bias_correcting_CMIP6_data.ipynb:
```diff
@@ -0,0 +1,1788 @@
+{
```
<p>"from the climate model simulations." instead of "climate model data" to avoid repeating data.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

richardarsenault replied:
<p>thanks! corrected.</p>

huard commented on docs/source/notebooks/08_Getting_and_bias_correcting_CMIP6_data.ipynb:
```diff
@@ -0,0 +1,1788 @@
+{
```
<p>I think in general it would be preferable to use average_shape instead of subset_shape, to account for variations in grid cell area across latitude.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

richardarsenault replied:
<p>I agree this would be best, but I can't find the method in Clisops? Maybe another package has this?</p>

richardarsenault replied:
Damn, I was importing only subset from clisops.core. That is why I couldn't find it...

richardarsenault replied:
<p>Changed for average_shape. </p>

huard commented on docs/source/notebooks/08_Getting_and_bias_correcting_CMIP6_data.ipynb:
```diff
@@ -0,0 +1,1788 @@
+{
```
<p>I suggest using a loop to avoid repeating the same code.</p><pre class="ql-syntax" spellcheck="false">out = []
for exp in ['historical', 'ssp585']:
    out[exp] = {}
    for variable in ['tasmin', 'tasmax', 'pr']:
        out[exp][variable] = ...
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='18'/>

richardarsenault replied:
<p>This returns an error, and I am unsure of what exactly the code is doing. Basically putting all the data into a 2D array of exp and variable?</p><p> </p><p>This is the code I wrote according to the comment above:</p><p> </p><pre class="ql-syntax" spellcheck="false">out = []
&nbsp;&nbsp;for exp in ['historical','ssp585']:
&nbsp;&nbsp;&nbsp;&nbsp;out[exp] = {}
&nbsp;&nbsp;&nbsp;&nbsp;for variable in ['tasmin','tasmax','pr']:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;query = dict(experiment_id=exp, table_id='day', variable_id=variable, member_id='r1i1p1f1', source_id=climate_model)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;col_subset = col.search(require_all_on=['source_id'], **query)
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;mapper = fsCMIP.get_mapper(col_subset.df.zstore[0])
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;out[exp][variable] = historical_tasmin=subset.subset_shape(xr.open_zarr(mapper, consolidated=True).sel(time=slice(reference_start_day,reference_end_day)), basin_contour).mean({'lat','lon'}).reset_coords('height',drop=True)[variable].chunk(-1)
</pre><p> </p><p> </p><p>This is the error I get:</p><p> </p><p> </p><pre class="ql-syntax" spellcheck="false">---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [15], in &lt;cell line: 2&gt;()
      5 out = []
      6 for exp in ['historical','ssp585']:
----&gt; 7     out[exp] = {}
      8     for variable in ['tasmin','tasmax','pr']:
      9         query = dict(experiment_id=exp, table_id='day', variable_id=variable, member_id='r1i1p1f1', source_id=climate_model)

TypeError: list indices must be integers or slices, not str
</pre><p> </p><p> </p><p>I also tried changing the curly braces for brackets, same error:</p><p> </p><pre class="ql-syntax" spellcheck="false">---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [16], in &lt;cell line: 2&gt;()
      5 out = []
      6 for exp in ['historical','ssp585']:
----&gt; 7     out[exp] = []
      8     for variable in ['tasmin','tasmax','pr']:
      9         query = dict(experiment_id=exp, table_id='day', variable_id=variable, member_id='r1i1p1f1', source_id=climate_model)

TypeError: list indices must be integers or slices, not str
</pre><p> </p><p>Any insights?</p>

huard replied:
<p>oups, start with <code>out={}</code></p>

richardarsenault replied:
<p>Thanks, works now!</p>

huard commented on docs/source/notebooks/08_Getting_and_bias_correcting_CMIP6_data.ipynb:
```diff
@@ -0,0 +1,1788 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #20.    print(corrected_ref_precip)
</pre><p>?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

richardarsenault replied:
<p>Vestige de debug... removed!</p>

huard commented on docs/source/notebooks/08_Getting_and_bias_correcting_CMIP6_data.ipynb:
```diff
@@ -0,0 +1,1788 @@
+{
```
<p>There is a strong annual cycle in the corrected dataset that isn't at all in the simulations. That's expected?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='27'/>

richardarsenault replied:
<p>The corrected version looks A LOT more like the ERA5 reference data, so I would say the method is correcting for this, yes. Now it remains to be seen if the ERA5 data is actually correct. I know @sebastienlanglois was looking into some possible artifacts. </p>

huard replied:
<p>Good for now. </p>

sebastienlanglois replied:
@richardarsenault  Indeed, please look into https://github.com/Ouranosinc/raven/issues/396#issuecomment-1140624233 for further info!

sebastienlanglois replied:
Also, I'm not a reviewer but was wondering why .to_dataset_dict() is not used after querying the esm catalog instead of accessing files with get_mapper ? This would remove the need to instantiate a GCS File System.

huard commented on docs/source/notebooks/09_Hydrological_impacts_of_climate_change.ipynb:
```diff
@@ -4,7 +4,7 @@
    "cell_type": "markdown",
```
<p>Can we fix the max_temp&lt;min_temp warnings ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

richardarsenault replied:
<p>I added a filter to stop showing the warnings here.</p>

huard commented on docs/source/notebooks/02_Extract_geographical_watershed_properties.ipynb:
```diff
@@ -0,0 +1,767 @@
+{
```
<pre class="ql-syntax" spellcheck="false">Line #9.        k: "{} %".format(int((v / total) * 100 * 10 ** precision) / 10 ** precision)
</pre><p>Is this very different from the <code>np.around</code> function, or just python's <code>round</code> ^ </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

richardarsenault replied:
<p>No idea. I took this directly from either the tests or an existing notebook, but I have no idea what this is doing nor if I can replace with a call to "round". I figured that since it's boilerplate and won't be changed by users, might as well leave it as such?</p>

richardarsenault replied:
Fixed, now use np.round, thanks!

huard commented on docs/source/notebooks/02_Extract_geographical_watershed_properties.ipynb:
```diff
@@ -0,0 +1,767 @@
+{
```
<p>Probably good to add a legend with the name of the categories.<em> </em>If you feel fancy, use the same color code for the bars in the first graph as in the second graph.</p><p> </p><pre class="ql-syntax" spellcheck="false">from matplotlib.colors import Normalize
norm = Normalize()
norm.autoscale(unique[:-1])
cm = plt.cm.get_cmap("tab20")
plt.bar(unique[:-1], counts[:-1], colors=cm(norm(unique[:-1])))
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='16'/>

richardarsenault replied:
<p>After tinkering with this for a long time, I was able to add your code here but not the legend. Unless I fix the categories, the number of "unique" and "counts" can vary and will not fit with a fixed legend. So instead, I added a cell with an explanation of the categories and I "beautified" the output of land-use types and counts.</p>

huard commented on docs/source/notebooks/02_Extract_geographical_watershed_properties.ipynb:
```diff
@@ -0,0 +1,767 @@
+{
```
<p>It's already open with rasterio, no ?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

richardarsenault replied:
<p>Changed sentence to reflect this</p>

huard commented on docs/source/notebooks/03_Extracting_forcing_data.ipynb:
```diff
@@ -0,0 +1,442 @@
+{
```
<p>automatically</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

richardarsenault replied:
<p>fixed</p>

huard commented on docs/source/notebooks/04_Emulating_hydrological_models.ipynb:
```diff
@@ -4,7 +4,7 @@
    "cell_type": "markdown",
```
<p>tailored</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

richardarsenault replied:
<p>fixed!</p>

huard commented on docs/source/notebooks/04_Emulating_hydrological_models.ipynb:
```diff
@@ -4,7 +4,7 @@
    "cell_type": "markdown",
```
<p>by adding the <code>workdir</code> path... will be made available from the PAVICS Jupyter environment. The default <code>workdir</code> setting puts model outputs in the temporary directory <code>/tmp</code> , which is not visible from the file Jupyter file explorer. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

richardarsenault replied:
<p>Good precision, fixed.</p>

huard commented on docs/source/notebooks/04_Emulating_hydrological_models.ipynb:
```diff
@@ -4,7 +4,7 @@
    "cell_type": "markdown",
```
<pre class="ql-syntax" spellcheck="false">Line #1.    # This syntax will allow you to get the path to where the data is located on the server.
</pre><p>These are the paths to the output files. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

richardarsenault replied:
<p>fixed</p>

huard commented on docs/source/notebooks/04_Emulating_hydrological_models.ipynb:
```diff
@@ -4,7 +4,7 @@
    "cell_type": "markdown",
```
<pre class="ql-syntax" spellcheck="false">Line #1.    # This syntax will allow you yo actually load the data in memory and work with it directly. Note that there is no way to work with the zip-file directly,
</pre><p>These are the actual outputs loaded as Python objects. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='21'/>

richardarsenault replied:
<p>true, fixed!</p>

huard commented on docs/source/notebooks/04_Emulating_hydrological_models.ipynb:
```diff
@@ -4,7 +4,7 @@
    "cell_type": "markdown",
```
<p>has</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Ouranosinc/raven/pull/424/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

richardarsenault replied:
<p>fixed</p>

richardarsenault commented on docs/source/notebooks/02_Extract_geographical_watershed_properties.ipynb:
```diff
@@ -0,0 +1,767 @@
+{
```
<p>fixed</p>

richardarsenault commented on docs/source/notebooks/04_Emulating_hydrological_models.ipynb:
```diff
@@ -4,7 +4,7 @@
    "cell_type": "markdown",
```
<p>Changed to "we can inspect and work with other state variables..."</p>

richardarsenault commented:
@huard I have updated the notebooks following your comments, but there is a new problem that I've never had before (not related to changes, it is now erroring even with previous notebooks). It's in Notebook 08 - Getting and bias-correcting CMIP6 data. 

Everything seems to work smoothly, but when comes the time to actually write the data to a netcdf file at the end of the chain, I get a "FSTimeoutError". I am assuming that this is a File System Timeout, so perhaps writing the files is taking too much time vs what is allowed in the FS config? It used to work without a hitch, but now it's erroring pretty quickly, within about a minute I would estimate.

Also, one comment regarding the legend of the land-use graphs was not completed in the same way as you expected, but I think the solution is OK (see NB 02_extrac_geographical_watershed...). Finally, one unresolved comment above regarding the rounding I am unsure of, I think probably @Zeitsperre coded this and I don't understand why, but I assume there is a reason!

Thanks!

richardarsenault commented:
Just added the notebooks with recent changes to the way ERA5 data is called, and fixed a bug in which future period data always has 0 timesteps. Also fixed the rounding function that was a bit convoluted.

richardarsenault commented:
It seems that out of nowhere, notebook 09 now fails with a weird error:

```
---------------------------------------------------------------------------
RavenError                                Traceback (most recent call last)
Input In [4], in <cell line: 23>()
     13 common_run_parameters = dict(
     14     area=4250.6,
     15     elevation=843.0,
   (...)
     19     rain_snow_fraction="RAINSNOW_DINGMAN", 
     20 )
     22 # Run the hydrological model using our forcing data from the reference period, and with the common parameters that will also be used for the future period simulation.
---> 23 model(
     24     forcing,
     25     start_date=dt.datetime(1981,1,1),
     26     end_date=dt.datetime(1990,12,31),
     27     **common_run_parameters,
     28      )
     30 # Now plot the results
     31 model.hydrograph.q_sim.plot()

File /opt/conda/envs/birdy/lib/python3.8/site-packages/ravenpy/models/base.py:431, in Raven.__call__(self, ts, overwrite, **kwds)
    428 messages = self.extract_raven_messages()
    430 if messages["ERROR"]:
--> 431     raise RavenError("\n".join(messages["ERROR"]))
    433 for msg in messages["WARNING"]:
    434     warn(msg, category=RavenWarning)

RavenError: NetCDF error [NetCDF: Attempt to convert between text & numbers] occured.
```

This was not the case recently, and I can't seem to get this working anymore. Perhaps a change in Raven or in the netcdf package installed on the server? This is all through pavics.ouranos.ca.

@huard @tlvu have any changes been made recently? thanks!

tlvu commented:
> @huard @tlvu have any changes been made recently? thanks!

@richardarsenault On May 5, we went live with the current Jupyter image, see all changes here https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/pull/105.

Can you try the "previous" Jupyter image on PAVICS?  That "previous" image was this PR https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/pull/101 and went live on Apr 7.

An even older Jupyter image is this one https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/pull/95, went live on Jan 27, not available on PAVICS anymore, if you need to try it, let me know and I'll try to work something out, probably giving you access to my dev VM.

Also, to future proof this PR, you might also want to try it on the "beta" image on PAVICS that I hope to be able to go live in 2, 3 weeks.  It has an even newer xclim.
